pkg/custom-plugins/spec_component.go

```diff
@@ -0,0 +1,211 @@
+package customplugins
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	apiv1 "github.com/leptonai/gpud/api/v1"
+	"github.com/leptonai/gpud/components"
+	"github.com/leptonai/gpud/pkg/log"
+)
+
+// NewInitFunc creates a new component initializer for the given plugin spec.
+func (spec *Spec) NewInitFunc() components.InitFunc {
+	if spec == nil {
+		return nil
+	}
+	return func(gpudInstance *components.GPUdInstance) (components.Component, error) {
+		cctx, ccancel := context.WithCancel(gpudInstance.RootCtx)
+		c := &component{
+			ctx:    cctx,
+			cancel: ccancel,
+			spec:   spec,
+		}
+		return c, nil
+	}
+}
+
+var _ components.Component = &component{}
+
+type component struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	spec *Spec
+
+	lastMu          sync.RWMutex
+	lastCheckResult *checkResult
+}
+
+var _ CustomPluginRegisteree = &component{}
+
+func (c *component) IsCustomPlugin() bool {
+	return true
+}
+
+func (c *component) Spec() Spec {
+	if c == nil || c.spec == nil {
+		return Spec{}
+	}
+	return *c.spec
+}
+
+var _ components.Deregisterable = &component{}
+
+func (c *component) CanDeregister() bool {
+	return true
+}
+
+func (c *component) Name() string { return c.spec.ComponentName() }
+
+func (c *component) Start() error {
+	log.Logger.Infow("starting custom plugin", "plugin", c.spec.PluginName)
+
+	itv := c.spec.Interval.Duration
+	// either periodic check is disabled or interval is too short
+	if itv < time.Second {
+		_ = c.Check()
+		return nil
+	}
+
+	go func() {
+		ticker := time.NewTicker(itv)
+		defer ticker.Stop()
+
+		for {
+			_ = c.Check()
+
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-ticker.C:
+			}
+		}
+	}()
+	return nil
+}
+
+func (c *component) Check() components.CheckResult {
+	log.Logger.Infow("checking custom plugin", "plugin", c.spec.PluginName)
+
+	cr := &checkResult{
+		ts: time.Now().UTC(),
+	}
+	defer func() {
+		c.lastMu.Lock()
+		c.lastCheckResult = cr
+		c.lastMu.Unlock()
+	}()
+
+	if c.spec.StatePlugin == nil {
+		cr.health = apiv1.HealthStateTypeHealthy
+		cr.reason = "no state plugin defined"
+		return cr
+	}
+
+	cctx, ccancel := context.WithTimeout(c.ctx, c.spec.Timeout.Duration)
+	defer ccancel()
+
+	cr.Output, cr.ExitCode, cr.err = c.spec.StatePlugin.executeAllSteps(cctx)
+
+	if cr.err != nil {
+		cr.health = apiv1.HealthStateTypeUnhealthy
+		cr.reason = fmt.Sprintf("error executing state plugin -- %s", cr.err)
+		return cr
+	}
+
+	cr.health = apiv1.HealthStateTypeHealthy
+	cr.reason = string(cr.Output)
```

Comment:
if it's healthy, we would like to not record the output unless we can limit the maximum length of the output.

Reply:
done